src/jit/namedintrinsiclist.h

```diff
@@ -16,10 +16,84 @@ enum NamedIntrinsic : unsigned int
     NI_System_Collections_Generic_EqualityComparer_get_Default = 4,
 #if FEATURE_HW_INTRINSICS
     NI_HW_INTRINSIC_START,
-#define HARDWARE_INTRINSIC(id, name, isa) NI_##id,
+#define HARDWARE_INTRINSIC(id, name, isa, ival, size, numarg, t1, t2, t3, t4, t5, t6, t7, t8, t9, t10, category)       \
+    NI_##id,
 #include "hwintrinsiclistxarch.h"
     NI_HW_INTRINSIC_END
 #endif
 };
 
+#if FEATURE_HW_INTRINSICS && defined(_TARGET_XARCH_)
+enum HWIntrinsicCategory : unsigned int
+{
+    // Simple SIMD intrinsics
+    // - take Vector128/256<T> parameters
+    // - return a Vector128/256<T>
+    // - generate single instruction
+    // - the codegen of overloads can be determined by intrinsicID and base type of returned vector
+    HW_Category_SimpleSIMD = 0x1,
+
+    // Scalar intrinsics
+    // - operate over general purpose registers, like crc32, lzcnt, popcnt, etc.
+    HW_Category_Scalar = 0x2,
+
+    // 64bit:
+    // - the generated instruction may require 64-bit general purpose register (r64)
+    // - must throw PlatforNotSupportedException on 32-bit platform
+    HW_Category_64bit       = 0x4,
+    HW_Category_64bitScalar = 0x6,
+
+    // IMM intrinsics
+    // - the generated instruction requires immediate value (i.e. imm8)
+    HW_Category_IMM = 0x8,
+
+    // Fixed IMM intrinsics
+    // - the immediate value is from ival field of the table, such as Sse.CompareEqual
+    HW_Category_FixedIMM = 0x10,
+
+    // Full range IMM intrinsic
+    // - the immediate value is vaild on the full range of imm8 (0-255)
+    HW_Category_FullRangeIMM = 0x20,
+
+    // Memory access intrinsics
+    // - e.g., Avx.Load, Avx.Store, Sse.LoadAligned
+    HW_Category_MemoryLoad  = 0x40,
+    HW_Category_MemoryStore = 0x80,
+
+    // Complex overload intrinsics
+    // - the codegen of overloads cannot be determined by intrinsicID and base type
+    HW_Category_ComplexOverloads = 0x100,
+
+    // Helper intrinsics
+    // - do not directly correspond to a instruction, such as Avx.SetAllVector256
+    HW_Category_Helper = 0x200,
```

Comment:
I think we should have a `IsHelper` flag and a `HasCodeGen` flag.

You can have a helper and it can either have codegen or not (`Set` is a helper without codegen, `SetZero` is a helper with codegen).

Reply:
So here I defined `HW_Category_NoCodeGenHelper ` for `Set ` and `HW_Category_Helper` for `SetZero`.